src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs

```diff
@@ -0,0 +1,138 @@
+            var targetFileName = state.TargetFileNameCandidate + state.TargetFileExtension;
+            var uiRequired = ProjectContainsTargetFile(state.SemanticDocument.Project, targetFileName, state.DocumentName, state.TypeName);
+            var isPartial = IsPartial(state.TypeNode);
+            var isNestedType = IsNestedType(state.TypeNode);
+            var singleType = IsSingleTypeDeclarationInSourceDocument(state.SemanticDocument.Root);
+            var typeSymbol = state.SemanticDocument.SemanticModel.GetDeclaredSymbol(state.TypeNode, cancellationToken) as INamedTypeSymbol;
+
+            if (singleType)
+            {
+                // one type declaration in current document. No moving around required, just sync
+                // document name and type name by offering rename in both directions between type and document.
+                AddSimpleCodeAction(actions, state, uiRequired, renameFile: true);
+                AddSimpleCodeAction(actions, state, uiRequired, renameType: true);
+            }
+            else
+            {
+                // multiple type declarations in current document.
+                AddSimpleCodeAction(actions, state, uiRequired, newFile: true, makeOuterTypePartial: isNestedType);
+            }
+
+            // Add actions that bring up UI dialog.
+            AddCodeActionsWithUI(actions, state, makeTypePartial: !isPartial, makeOuterTypePartial: isNestedType);
+
+            return actions;
+        }
+
+        private void AddCodeActionsWithUI(
+            List<CodeAction> actions,
+            State state,
+            bool makeTypePartial,
+            bool makeOuterTypePartial)
+        {
+            // make this type declaration partial and add a new partial part in a new file.
+            if (makeTypePartial)
+            {
+                actions.Add(
+                    GetCodeActionWithUI(state, makeTypePartial, makeOuterTypePartial));
+            }
+
+            // this gives an option for the user specify a custom file name for the normal move scenario.
+            actions.Add(
+                GetCodeActionWithUI(
+                    state, makeTypePartial: false, makeOuterTypePartial: makeOuterTypePartial));
+        }
+
+        private MoveTypeCodeActionWithOption GetCodeActionWithUI(
+            State state, bool makeTypePartial, bool makeOuterTypePartial) =>
+                new MoveTypeCodeActionWithOption(
+                        (TService)this,
+                        state: state,
+                        makeTypePartial: makeTypePartial,
+                        makeOuterTypePartial: makeOuterTypePartial);
+
+        private void AddSimpleCodeAction(
+            List<CodeAction> actions,
+            State state,
+            bool uiRequired,
+            bool renameFile = false,
+            bool renameType = false,
+            bool newFile = false,
+            bool makeTypePartial = false,
+            bool makeOuterTypePartial = false)
```

Comment:
Isn't makeOuterTypePartial unnecessary?  Can't it be inferred by the operatoin you're performing?